File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.py
from collections import deque
import sys
input = sys.stdin.readline

def bfs(start):
    q = deque()
    q.append(start)
    visited[start] = True
    while q:
        cur = q.popleft()
        for i in range(1,7):
            ncur = cur+i
            if ncur==100:
                board[ncur]=board[cur]+1
                visited[ncur] = True
                return
            if 0<ncur<=100 and not visited[ncur]:
                if ncur in ladder:
                    visited[ncur] = True
                    board[ncur] = board[cur]+1

                    if not visited[ladder[ncur]]:
                        visited[ladder[ncur]]=True
                        board[ladder[ncur]] = board[ncur]
                    else:
                        board[ladder[ncur]]=min(board[ladder[ncur]], board[ncur])

                    q.append(ladder[ncur])
                    
                elif ncur in snake:
                    visited[ncur] = True
                    board[ncur] = board[cur]+1

                    if not visited[snake[ncur]]:
                        visited[snake[ncur]]=True
                        board[snake[ncur]] = board[ncur]
                    else:
                        board[snake[ncur]]=min(board[snake[ncur]], board[ncur])

                    q.append(snake[ncur])

                else:
                    q.append(ncur)
                    visited[ncur] = True
                    board[ncur] = board[cur]+1
                

N,M = map(int,input().split())
board = [0]*101
visited= [False]*101
ladder = {}
snake = {}

for _ in range(N):
    x,y = map(int,input().split())
    ladder[x] = y
for _ in range(M):
    u,v = map(int,input().split())
    snake[u] = v

bfs(1)
print(board[100])

File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.java
import java.io.*;
import java.util.*;

public class Main {
    static int N;
    static int M;
    static boolean[] visited;
    static int[] steps;
    
    static HashMap<Integer, Integer> ladders = new HashMap<Integer, Integer>();
    
    static void bfs(int start){
        Deque<Integer> q = new ArrayDeque<Integer>();
        q.add(start);
        visited[start] = true;
        
        while(!q.isEmpty()){
            int cur = q.poll();
            
            for(int i=1;i<=6;i++){
                int next = cur+i;
                
                if(next==100){
                    steps[next] = steps[cur]+1;
                    visited[next] = true;
                    return;
                }
                
                if(next <= 100 && !visited[next]){
                    visited[next] = true;
                    steps[next] = steps[cur]+1;
                    
                    if(ladders.get(next) != null){
                        int ladderNext = ladders.get(next);
                        
                        if(!visited[ladderNext]){
                            steps[ladderNext] = steps[next];
                            visited[ladderNext] = true;
                        }else{
                            steps[ladderNext] = Math.min(steps[ladderNext], steps[next]);
                        }
                        
                        q.add(ladderNext);
                    }
                    
                    else {
                        q.add(next);
                    }
                    
                }
            }
        }
    }
    
  public static void main(String args[]) throws IOException{
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    
    StringTokenizer st = new StringTokenizer(br.readLine());
    
    N = Integer.parseInt(st.nextToken());
    M = Integer.parseInt(st.nextToken());
    
    visited = new boolean[101];
    steps = new int[101];
    
    
    for(int i=0;i<N+M;i++){
        st = new StringTokenizer(br.readLine());
        ladders.put(Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken()));
    }
    
    bfs(1);
    System.out.println(steps[100]);
  }
}